docs/langref/relay_expr.rst

```diff
@@ -0,0 +1,462 @@
+using the variables repeatedly, as below (a C++ program using the corresponding API bindings
+could accomplish the same thing):
+
+.. code-block:: python
+
+   sum1 = relay.add(a, b)
+   sum2 = relay.add(sum1, sum1)
+   relay.multiply(sum2, sum2)
+
+For development purposes and to enable certain optimizations, Relay includes passes to
+convert between dataflow graphs defined using graph bindings and programs with :code:`let`
+bindings in A-normal form, employed by many compiler optimizations from the functional
+programming community (see `"The Essence of Compiling with Continuations" by
+Flanagan *et al*<https://slang.soe.ucsc.edu/cormac/papers/pldi93.pdf>`__ for a discussion
+of the A-normal form).
+
+=======================
+Control Flow Expression
+=======================
+
+Control flow expressions allow the network topology to change
+based on the value of previously executed expressions.
+
+Call
+~~~~
+
+Expressions with function types in Relay are "callable,"
+meaning that they can be invoked via a function call. These consist of
+any expression that evaluates to a closure (i.e., function expressions
+or global functions) and Relay operators.
+
+When a closure is called, the body is evaluated in the closure's stored
+environment (i.e., using the stored values for free variables) with
+local variable bindings added for each argument; the final value
+obtained by evaluating the body is the call's return value.
+In the case of operators, the implementation is opaque to Relay,
+so the result is left up to the registered TVM implementation.
+
+For example, we can call the previously defined `%fact` because it
+has a function type:
+
+.. code-block:: python
+
+   %fact(10)
+
+A type-polymorphic function can also include type arguments at a call
+site. The type arguments are substituted for type parameters when
+type checking.
+
+Note that all type relations in the function type must hold at each
+call site. Because relations are checked at call sites, this means that
+the relations are checked against the types of the particular arguments
+at that call site, so this also allows for a form of polymorphism.
+
+See :py:class:`~tvm.relay.expr.Call` for its definition and documentation.
+
+If-Then-Else
+~~~~~~~~~~~~
+
+Relay has a simple if-then-else expression that allows programs to branch
+on a single value of type :code:`bool`, i.e., a zero-rank
+tensor of booleans (:code:`Tensor[(), bool]`).
+
+.. code-block:: python
+
+    if (equal(%t, %u)) {
+        %t
+    } else {
+        %u
+    }
+
+Since if-then-else branches are expressions, they may appear inline
+wherever any other expression may be expected, like invocations of
+the ternary operator in C-like languages. The if-then-else expression
+evaluates to the value of the "then" branch if the condition value is
+`True` and evaluates to the value of the "else" branch otherwise.
+
+See :py:class:`~tvm.relay.expr.If` for its definition and documentation.
```

Comment:
Should there be a discussion of TempExpr?

Reply:
TempExpr are a purely internal implementation detail, so I don't know if they need to be in the language reference either.